### PR TITLE
Package based targets for enrichments

### PR DESCRIPTION
**Pull request by dmdashenkov**
#### Example:

```
message EnrichmentForAllEventsInPackage {
     option (enrichment_for) = "com.foo.bar.*";
     ...
}
```

_Note:_ no changes in the `tools` module are required.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol PTAL.

File: client/src/main/java/org/spine3/protobuf/KnownTypes.java [hunk 1]
```diff
@@ -161,6 +164,23 @@ public static TypeUrl getTypeUrl(String typeName) {
         return typeUrl;
     }
 
+    public static Collection<TypeUrl> getTypesFromPackage(final String packageName) {
```
Comment (armiol) [inline client/src/main/java/org/spine3/protobuf/KnownTypes.java]: `Set`?

File: client/src/test/proto/spine/test/wildcard_messages.proto [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+syntax = "proto3";
+
+package spine.test.wildcard;
```
Comment (armiol) [inline client/src/test/proto/spine/test/wildcard_messages.proto]: `spine.test`.

Comment (armiol) [inline client/src/test/proto/spine/test/wildcard_messages.proto]: As long as this file belongs to `known_types`, please name it accordingly. And leave a comment in the file.

File: server/src/test/proto/spine/test/event/enrichment/otherpackage/events.proto [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+syntax = "proto3";
+
+package spine.test.event.enrichment.otherpackage;
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/enrichment/otherpackage/events.proto]: Let's move it to the `spine.test.event.user` (both package and folder).

And rename the events to belong to the `Users` domain.

File: server/src/test/proto/spine/test/event/enrichment/enrichment.proto [hunk 4]
```diff
@@ -43,3 +44,9 @@ message ProjectCreatedEnrichmentAnotherPackageFqnAndMsgOpt {
 
     string project_name = 1 [(by) = "spine.test.event.ProjectCreated.project_id"];
 }
+
+message ProjectPackageEventsEnrichment {
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/enrichment/enrichment.proto]: Let's rename it so it belongs to the `Users` domain.

File: server/src/test/proto/spine/test/event/enrichment/enrichment.proto [hunk 5]
```diff
@@ -43,3 +44,9 @@ message ProjectCreatedEnrichmentAnotherPackageFqnAndMsgOpt {
 
     string project_name = 1 [(by) = "spine.test.event.ProjectCreated.project_id"];
 }
+
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/enrichment/enrichment.proto]: Let's add a comment, which package is a target. Plus, I would mention the name of the unit test.

File: server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java [hunk 6]
```diff
@@ -92,8 +100,33 @@ private void put(Properties props) {
 
         private void put(String enrichmentType, Iterable<String> eventTypes) {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java]: As discussed, let's rename it to `eventQualifier`.

File: server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java [hunk 7]
```diff
@@ -92,8 +100,33 @@ private void put(Properties props) {
...
             for (String eventType : eventTypes) {
-                builder.put(enrichmentType, eventType);
+                if (isPackage(eventType)) {
+                    putAll(enrichmentType, eventType);
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java]: `putAllTypesFromPackage`

File: server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java [hunk 8]
```diff
@@ -92,8 +100,33 @@ private void put(Properties props) {
...
+                } else {
+                    builder.put(enrichmentType, eventType);
+                }
+            }
+        }
+
+        private void putAll(String enrichmentType, String eventsPackage) {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java]: Let's add a small JavaDoc comment explaining the reasoning.

File: server/src/test/proto/spine/test/event/enrichment/enrichment.proto (hunk 5 above)
Comment (dmdashenkov) [inline server/src/test/proto/spine/test/event/enrichment/enrichment.proto]: Done

**Comment by dmdashenkov** [comment]
@armiol PTAL again

**Comment by dmdashenkov** [comment]
@armiol BTW, should we bump the version for this change?

File: server/src/test/proto/spine/test/event/enrichment/enrichment.proto [hunk 9]
```diff
@@ -43,3 +44,11 @@ message ProjectCreatedEnrichmentAnotherPackageFqnAndMsgOpt {
...
+// Enrichment for all the event in the "spine.test.event.user" package.
+// See {@code EventEnrichmentsMapShould#contain_all_events_from_package_by_one_enrichment} test.
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/enrichment/enrichment.proto]: Let's simplify to 
```
// The enrichment for all the events declared in the "spine.test.event.user" package.
//
// See `EventEnrichmentsMapShould#contain_all_events_from_package_by_one_enrichment` test for details.
//
```

**Comment by armiol** [comment]
@dmdashenkov yes, let's bump the version.

**Comment by dmdashenkov** [comment]
Latest test cases added:
 - multiple package enrichment;
 - package and single event enrichment;
 - sub-pacakges;
 - missing target property events.

@armiol PTAL

File: client/src/main/java/org/spine3/protobuf/KnownTypes.java [hunk 10]
```diff
@@ -161,6 +172,47 @@ public static TypeUrl getTypeUrl(String typeName) {
         return typeUrl;
     }
 
+    public static Set<TypeUrl> getTypesFromPackage(final String packageName) {
```
Comment (armiol) [inline client/src/main/java/org/spine3/protobuf/KnownTypes.java]: Please add some JavaDocs for the `public static ...` methods in this class.

File: client/src/test/proto/spine/test/types/known_types_should.proto [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+option java_package="org.spine3.test.types";
+option java_multiple_files = true;
+
+import "spine/annotations.proto";
+
+// The types declared below are used to test the `KnownTypes#getTypesFromPackage`.
+//
+// See `KnownTypesShould` for the unit test code.
+//
+
```
Comment (armiol) [inline client/src/test/proto/spine/test/types/known_types_should.proto]: Please remove an empty line.

File: server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java [hunk 12]
```diff
@@ -66,6 +81,13 @@ private EventEnrichmentsMap() {}
 
     private static class Builder {
 
+        /**
+         * Constant indicating a package qualifier.
+         */
+        private static final String PACKAGE_WILDCARD_INDICATOR = ".*";
+
+        private static final String WILDCARD_TYPE_UPON_FIELD_INDICATOR = "*.";
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java]: The name is not clear enough. Please add a documentation for it. Maybe you'll find a better name while writing the docs.

File: server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java [hunk 13]
```diff
@@ -85,15 +107,84 @@ private void put(Properties props) {
...
+            final Descriptor enrichmentDescriptor = KnownTypes.getDescriptorForType(enrichmentType);
+
+            final Set<String> result = new HashSet<>();
+            for (FieldDescriptor field : enrichmentDescriptor.getFields()) {
+                final String extension = field.getOptions()
+                                              .getExtension(EventAnnotationsProto.by);
+                final String fieldName = extension.substring(WILDCARD_TYPE_UPON_FIELD_INDICATOR.length());
+                result.add(fieldName);
             }
+
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/enrich/EventEnrichmentsMap.java]: Please remove an empty line.

File: server/src/test/proto/spine/test/event/enrichment/enrichment.proto [hunk 14]
```diff
@@ -43,3 +45,46 @@ message ProjectCreatedEnrichmentAnotherPackageFqnAndMsgOpt {
...
+
+// The enrichment for all the event declared in "spine.test.event.user.permissions" package and referencing user who
+// is able to grant permissions.
+//
+// See `EventEnrichmentsMapShould#contain_only_events_with_target_field_if_declared_though_package` test for details.
+//
+message GranterEventsEnrichment {
+    option (enrichment_for) = "spine.test.event.user.permissions.*";
+
+    spine.people.PersonName granter_name = 1 [(by) = "*.granter_uid"];
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/enrichment/enrichment.proto]: There is no such a word as `granter`. The correct term is `grantor`, i.e. the person who grants.

File: server/src/test/proto/spine/test/event/user/connections/events.proto [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+syntax = "proto3";
+
+package spine.test.event.user.connections;
+
+option (type_url_prefix) = "type.spine3.org";
+option java_package="org.spine3.test.event.user.connections";
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/user/connections/events.proto]: Usually we don't use plural in package names. Please revisit.

File: server/src/test/proto/spine/test/event/user/connections/events.proto [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+option java_multiple_files = true;
+
+import "spine/annotations.proto";
+import "spine/enrichment_annotations.proto";
+
+message UserConnectionEstablished {
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/user/connections/events.proto]: I would rethink the domain. `Connection` is too generic and unclear.

Let's transform it into something like `SharingRequestSent`, turning the next event into `SharingRequestApproved`.

File: server/src/test/proto/spine/test/event/user/permissions/events.proto [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import "spine/annotations.proto";
+import "spine/enrichment_annotations.proto";
+
+message PermissionGrantedEvent {
+
+    string user_uid = 1;
+    string permission_id = 2;
+
+    // ID of the user who granted the permission
+    string granter_uid = 3;
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/user/permissions/events.proto]: Please rename. See above.

File: server/src/test/proto/spine/test/event/user/permissions/events.proto [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+syntax = "proto3";
+
+package spine.test.event.user.permissions;
+
+option (type_url_prefix) = "type.spine3.org";
+option java_package="org.spine3.test.event.user.permissions";
```
Comment (armiol) [inline server/src/test/proto/spine/test/event/user/permissions/events.proto]: Again, please use singular for the package name.

**Comment by dmdashenkov** [comment]
@armiol PTAL again.

**Pull request merged**